src/ReaderThemeSupportFeatures.php

```diff
@@ -0,0 +1,313 @@
+
+	/**
+	 * Theme feature slug for editor-color-palette.
+	 *
+	 * @var string
+	 */
+	const FEATURE_EDITOR_COLOR_PALETTE = 'editor-color-palette';
+
+	/**
+	 * Theme feature slug for editor-gradient-presets.
+	 *
+	 * @var string
+	 */
+	const FEATURE_EDITOR_GRADIENT_PRESETS = 'editor-gradient-presets';
+
+	/**
+	 * Theme feature slug for editor-font-sizes.
+	 *
+	 * @var string
+	 */
+	const FEATURE_EDITOR_FONT_SIZES = 'editor-font-sizes';
+
+	/**
+	 * Supported features.
+	 *
+	 * @var string[]
+	 */
+	const SUPPORTED_FEATURES = [
+		self::FEATURE_EDITOR_COLOR_PALETTE,
+		self::FEATURE_EDITOR_GRADIENT_PRESETS,
+		self::FEATURE_EDITOR_FONT_SIZES,
+	];
+
+	/**
+	 * Reader theme loader.
+	 *
+	 * @var ReaderThemeLoader
+	 */
+	private $reader_theme_loader;
+
+	/**
+	 * ReaderThemeLoader constructor.
+	 *
+	 * @param ReaderThemeLoader $reader_theme_loader Reader theme loader.
+	 */
+	public function __construct( ReaderThemeLoader $reader_theme_loader ) {
+		$this->reader_theme_loader = $reader_theme_loader;
+	}
+
+	/**
+	 * Register the service with the system.
+	 *
+	 * @return void
+	 */
+	public function register() {
+		add_filter( 'amp_options_updating', [ $this, 'filter_amp_options_updating' ] );
+		add_action( 'after_switch_theme', [ $this, 'update_cached_theme_support' ] );
+
+		add_action(
+			'after_setup_theme',
+			[ $this, 'add_primary_theme_support' ],
+			100 // After the theme has added its own theme support.
+		);
+
+		add_action( 'amp_post_template_head', [ $this, 'print_theme_support_styles' ] );
+		add_action(
+			'wp_head',
+			[ $this, 'print_theme_support_styles' ],
+			9 // Because wp_print_styles happens at priority 8, and we want the primary theme's colors to override any conflicting theme color assignments.
+		);
+
+		// @todo Also do this when the theme has been updated?
+	}
+
+	/**
+	 * Filter the AMP options when they are updated to add the primary theme's features.
+	 *
+	 * @param array $options Options.
+	 * @return array Options.
+	 */
+	public function filter_amp_options_updating( $options ) {
+		if ( $this->reader_theme_loader->is_enabled( $options ) ) {
+			$options[ Option::PRIMARY_THEME_SUPPORT ] = $this->get_theme_support_features();
+		} else {
+			$options[ Option::PRIMARY_THEME_SUPPORT ] = null;
+		}
+		return $options;
+	}
+
+	/**
+	 * Update primary theme's cached theme support.
+	 */
+	public function update_cached_theme_support() {
+		if ( $this->reader_theme_loader->is_enabled() ) {
+			AMP_Options_Manager::update_option( Option::PRIMARY_THEME_SUPPORT, $this->get_theme_support_features() );
+		} else {
+			AMP_Options_Manager::update_option( Option::PRIMARY_THEME_SUPPORT, null );
+		}
+	}
+
+	/**
+	 * Get the theme support features.
+	 *
+	 * @return array Theme support features.
+	 */
+	private function get_theme_support_features() {
+		$features = [];
+		foreach ( self::SUPPORTED_FEATURES as $feature_key ) {
+			$features[ $feature_key ] = current( (array) get_theme_support( $feature_key ) );
```

Comment:
We can limit the keys that we store to only include `slug`, `size`, `color`, and `gradient`.